### ISSUE TITLE
Concerning performance with one-to-many query

### ISSUE THREAD

**Issue by ra50** [prisma/prisma#16889]
### Bug description

I'm doing a very simple PostgreSQL query across two tables (A, B) with a one-to-many relationship (A has many B) that returns 4 columns and about 7500 rows.  The query uses ILIKE to achieve case-insensitivity on table A with no index.

If I write this as an inner join by hand it looks like this:
```
SELECT col_A, col_B, col_c, col_D
FROM table_A a INNER JOIN table_B b on a.a_id = b.a_id
WHERE col_A ILIKE '%query%' or col_B ILIKE '%query%'
```
Even without the index this query runs in about 300 ms.

In Prisma, I wrote this as a `findMany` with `mode: 'insensitive'`.  I wish this resulted in the hand-written JOIN query above, but I know that Prisma executes it as two queries.  The first query returns about 4500 values that are used in an IN in the second query.

With Prisma these two queries (really, the second query with the IN) take 25 seconds to complete.  When running by hand, even these two separate queries don't take much longer than the single JOIN query above.  Something is happening when Prisma takes the 4500 results from the first query and puts it into the IN in the second query that makes something take 25 sec that should take < 1 sec.

To make sure it wasn't due to loading the data into memory, I tried taking the hand-written JOIN query and used it with `$queryRaw` and it also took less than 1 sec.

Is something wrong with the way the Prisma query is written?  I understand it's common to drop down to raw SQL for performance-sensitive things, but this isn't a very advanced query and it seems like Prisma shouldn't be taking 40x as long to do it.

### How to reproduce

1. Construct a one-to-many Prisma query similar to below on a large dataset (100-200k rows per table)
2. Compare time to retrieve results to 1) running a hand-written JOIN query and 2) running the two Prisma-generated queries by hand

### Expected behavior

A much smaller performance delta as opposed to 40-80x

### Prisma information

```prisma
// Add your schema.prisma
model TableA {
  aId         String         @id @map("a_id")
  colA       String?       @map("col_A")
  colB       String?       @map("col_B")
  tableBs TableB[]

  @@map("table_A")
}

model TableB {
  aId      String      @map("a_id")
  colC    String?    @map("col_C")
  colD    String?    @map("col_D")
  tableA TableA     @relation(fields: [aId], references: [aId])

  @@map("table_B")
}
```

```ts
// Add your code using Prisma Client
   prisma.tableA.findMany({
      where: {
        OR: [
          {
            colA: {
              contains: query,
              mode: 'insensitive'
            }
          },
          {
            colB: {
              contains: query,
              mode: 'insensitive'
            }
          }
        ]
      },
      select: {
        colA: true,
        colB: true,
        tableBs: {
          select: {
            colC: true,
            colD: true
          }
        }
      }
    })
```

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v16.13.1

### Prisma Version

```
4.3.1
```

**Comment by ra50** [comment]
Hi, is there enough detail provided to investigate this issue?

**Comment by janpio** [comment]
Generally it looks like that. You could also provide a reproduction repository we can just run out of the box, that would save us even more time when we finally get to this.

**Comment by janpio** [comment]
@ra50 Sorry we didn't follow up sooner. Did you maybe have time to create a reproduction repository for this? That would be the easiest way for us to understand the performance problem you are describing.

---

A note: The main query about JOINs usage (which would probably help here), where we will post updates and where a lot of discussion is happening right now is https://github.com/prisma/prisma/issues/5184. Make sure to follow that as well, although I will of course also try to update here if something relevant changes. Thanks.

**Comment by janpio** [comment]
I spent some time trying to reproduce what you are seeing @ra50. See my repository at https://github.com/janpio/16889

I took your [Prisma schema](https://github.com/janpio/16889/blob/main/prisma/schema.prisma) (needed to add an explicit `@id` to `tableB` to make it valid) and the [Prisma Client query](https://github.com/janpio/16889/blob/8bd9064f868ec5fe158a8c468f975b08ff3c4a39/script.js#L48-L75) you shared, also [the raw query](https://github.com/janpio/16889/blob/8bd9064f868ec5fe158a8c468f975b08ff3c4a39/script.js#L42), and turned them into a project. I also wrote a [small seed script](https://github.com/janpio/16889/blob/main/prisma/seed.js) that creates 5 entries in `tableA` and 4500 each in `tableB` - I wasn't super sure about your concrete number. So in total 22.500 entries - which are then all returned by both queries. 

<img width="669" alt="image" src="https://github.com/prisma/prisma/assets/183673/b7dd4275-8f98-43b1-9caa-b6339820a249">
<img width="766" alt="image" src="https://github.com/prisma/prisma/assets/183673/d2e514e2-5c78-4bb7-adab-f4948ec455d8">

This is typical representative measurement of what I get (both with Prisma 4.3.1 and Prisma 5.0.0):
```
> node .\script.js
$connect: 826.825ms
------------------
------------------
------------------
$queryRaw: 442.069ms
------------------
findMany: 532.211ms
------------------
$queryRaw: 482.847ms
------------------
findMany: 493.469ms
------------------
$queryRaw: 538.881ms
------------------
findMany: 566.284ms
------------------
$queryRaw: 535.71ms
------------------
findMany: 560.552ms
------------------
```

Note that I run both queries a few times, and skip the first output of each - Node and the database both are slower on the first time I run the queries - so that would give a wrong impression of what is going on.

So to me it looks that the raw query is usually a bit faster, but at the cost of returning a flat representation of the data while the Prisma Client query will already put it into a nice structure that matches the Prisma schema.

Did I miss something or do it wrong in some way that I get so different resulsts than what you describe?

**Comment by janpio** [comment]
Hey, @ra50, are you still around and could help me understand if I missed something in my reproduction attempt or if this is indeed fixed now? Thanks.

**Comment by ra50** [comment]
Hey @janpio, thank you for trying this!  Yes I'm still around, I'll update all the code and see if I can reproduce or why the results might be different.  I'll also follow the other thread.  I'll report back here and let you know.  Thanks.

**Comment by ra50** [comment]
Hi @janpio, I really appreciate you taking the time to recreate this since I didn't provide an example repo.  I looked at our code and we ended up removing these tables from our Prisma schema and using `$queryRaw`, this works for us since they're relatively static queries.  So it wouldn't make sense to try to recreate it at this time.  When I wrote this issue I tried comparing `$queryRaw` and the `findMany` back to back to make sure no application code was the cause.

It's good to see you didn't have the performance issue in your setup, and I'm interested in seeing what comes of future work on adding JOINs to the query engine.  We're still using Prisma in the rest of the codebase and are extremely pleased with it. 
 Thanks for your work on Prisma 5!
